internal/examples/server/data/agent.go

```diff
@@ -194,17 +194,35 @@ func (agent *Agent) updateEffectiveConfig(
 	}
 }
 
+func (agent *Agent) hasCapability(capability protobufs.AgentCapabilities) bool {
+	return agent.Status.Capabilities&capability != 0
+}
+
 func (agent *Agent) processStatusUpdate(
 	newStatus *protobufs.AgentToServer,
 	response *protobufs.ServerToAgent,
 ) {
-	if agent.Status != nil && agent.Status.SequenceNum+1 != newStatus.SequenceNum {
-		// We lost the previous status update. Request full status update from the agent.
-		response.Flags |= protobufs.ServerToAgent_ReportFullState
-	}
+	// We don't have any status for this Agent, or we lost the previous status update from the Agent, so our
+	// current status is not up-to-date.
+	lostPreviousUpdate := (agent.Status == nil) || (agent.Status != nil && agent.Status.SequenceNum+1 != newStatus.SequenceNum)
 
 	agentDescrChanged := agent.updateStatusField(newStatus)
 
+	// Check if any fields were omitted in the status report.
+	effectiveConfigOmitted := agent.hasCapability(protobufs.AgentCapabilities_ReportsEffectiveConfig) && newStatus.EffectiveConfig == nil
+	packageStatusesOmitted := agent.hasCapability(protobufs.AgentCapabilities_ReportsPackageStatuses) && newStatus.PackageStatuses == nil
+	remoteConfigStatusOmitted := agent.hasCapability(protobufs.AgentCapabilities_AcceptsRemoteConfig) && newStatus.RemoteConfigStatus == nil
+	healthOmitted := agent.hasCapability(protobufs.AgentCapabilities_ReportsHealth) && newStatus.Health == nil
+
+	// True if the status was not fully reported.
+	statusIsCompressed := effectiveConfigOmitted || packageStatusesOmitted || remoteConfigStatusOmitted || healthOmitted
+
+	if statusIsCompressed && lostPreviousUpdate {
```

Comment:
Should it be OR?

Reply:
No, "AND" is correct here. We only ned to ask for the full state if both of these conditions are true:
- The full state was not reported in the message we have just received. If this is not true, it means we have already just received the full state, there is no need to ask for it again.
- We detected that we lost a previous update. If this is not true, it means we already have the correct state on our end, there is no need to ask the client to send it again.

Reply:
Got it, thanks!